quantilizer.py

```diff
@@ -70,32 +73,102 @@ def get_trajectories(pi, env, horizon, n_trajectories):
-	return np.eye(27)[encoded_labels]
-
-def train(g_step = 5, max_iters = 1e5, adam_epsilon=1e-8, 
-		optim_batch_size = 256, reg = 1e-2, optim_stepsize = 3e-4, ckpt_dir = None , verbose=True, 
-		hidden_size = 20, reuse = False, horizon = 200, dataset_name='ryan', env_name='Hopper-v2', quantiles=[1.0, .5, .25, .125]):
+class ClassificationModel(object):
+	def __init__(self, nb_clf, input_dim, dataset_name,
+				env_name, q, framework='sklearn', reg=1e-4, hidden_size=20, aggregate_method='continuous', seed=0):
+		self.nb_clf = nb_clf
+		self.framework = framework
+		self.dataset_name = dataset_name
+		self.env_name = env_name
+		self.q = q
+		self.reg = reg
+		self.nb_model = nb_clf
+		self.input_dim = input_dim
+		self.classes = self.compute_classes()
+		self.hidden_size = hidden_size
+		self.seed = seed
+		self.model_list = self.init_models()
+		self.aggregate_method = aggregate_method
+		if not os.path.exists('log/models'):
+			os.makedirs('log/models')
+	def init_models(self):
+		if self.framework == 'keras':
+			# fix seeds to get reproducible results
+			np.random.seed(self.seed)
+			tf.set_random_seed(self.seed)
+			return [mlp_classification(self.input_dim, self.classes[i].shape[-1], reg=self.reg) for i in range(self.nb_model)]
+		elif self.framework == 'sklearn':
+			return [MLPClassifier(hidden_layer_sizes=(self.hidden_size, self.hidden_size), alpha=self.reg, random_state=self.seed) for _ in range(self.nb_model)]
+
+	def compute_classes(self):
+		"""gives the array of classes to predict for the quantile q dataset"""
+
+		filename = 'log/{}/{}.npz'.format(self.env_name, self.dataset_name)
+		dataset = Dataset(filename, quantile=self.q)
+		return [np.unique(dataset.acs[:,i]) for i in range(self.nb_clf)]
+	
+	def filename(self, index):
+		return 'log/models/{}_{}_{}_{}_{}_{}.h5'.format(self.dataset_name, self.env_name, self.q, index, self.framework, self.seed)
+
+	def fit(self, x_train, y_train):
+		for index, model in enumerate(self.model_list):
+			if self.framework == 'keras':
+				#model.fit(x_train, to_categorical(y_train[:,index] - y_train[:,index].min(), num_classes=self.classes[index].shape[-1]))
+				model.fit(x_train, y_train[:,index]-y_train[:,index].min())
+			elif self.framework == 'sklearn':
+				model.fit(x_train, y_train[:, index])
+				train_score = model.score(x_train, y_train[:, index])
+				print("train score q={}: {}".format(self.q, train_score))
+
+	def predict(self, x):
+		if self.framework == 'keras':
+				if self.aggregate_method == 'continuous':
+					return [(self.classes[index] * clf.predict(x.reshape(1, -1)).ravel()).sum() for (index,clf) in enumerate(self.model_list)]
+		elif self.framework == 'sklearn':
+			if self.aggregate_method == 'continuous':
+				return [(clf.classes_ * clf.predict_proba(x.reshape(1, -1)).ravel()).sum() for clf in self.model_list]
+			elif self.aggregate_method == 'argmax':
+				return [(clf.classes_ * clf.predict(x.reshape(1, -1)).ravel()).sum() for clf in self.model_list]
```

Comment:
does clf_predict output an (n,3) boolean array, or an (n,1) integer array? If the latter, then you're multiplying clf.classes_, which is np.array([0,1,2]) by an integer array, which confuses me.

Reply:
Here `clf.classes_` is `array([-1.0, 0.0, 1.0])`, `x` is of shape (11,) and `clf.predict_proba(x.reshape(1,-1))` is, for instance, `array([[p1, p2, 1-p1-p2]])`. To multiply with `clf.classes_` you apply .ravel() (i.e. you get array([p1, p2, 1-p1-p2])). The `.sum()` only happens after the multiplication with `clf.classes_`, and you get something like `-1*p1 + 0*p2 + 1*(1-p1-p2)`.